File: api/khairat-api/src/main/java/my/emasjid/khairatapi/service/DependentService.java
package my.emasjid.khairatapi.service;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;

import my.emasjid.khairatapi.entity.Dependent;
import my.emasjid.khairatapi.entity.Member;
import my.emasjid.khairatapi.entity.Person;
import my.emasjid.khairatapi.repository.DependentRepository;
import my.emasjid.khairatapi.repository.MemberRepository;
import my.emasjid.khairatapi.repository.PersonRepository;

@Service
public class DependentService {
    
    @Autowired
    private DependentRepository dependentRepository;
    @Autowired
    private PersonRepository personRepository;
    @Autowired
    private MemberRepository memberRepository;

    public Dependent save(Dependent dependent, Long memberId) {
        Person newPerson = dependent.getPerson();
        newPerson = personRepository.save(newPerson);
        dependent.setPerson(newPerson);
        Member member = memberRepository.findById(memberId).orElse(null);
        dependent.setMember(member);
        return dependentRepository.save(dependent);
    }

    public void deleteById(Long id) {
        Dependent dependent = dependentRepository.findById(id).orElse(null);
        Person person = personRepository.findById(dependent.getPerson().getId()).orElse(null);
        dependentRepository.delete(dependent);
        personRepository.delete(person);
    }
}

File: api/khairat-api-go/service/dependent_service.go
package service

import (
	"github.com/Dev4w4n/e-masjid.my/api/khairat-api/model"
	"github.com/Dev4w4n/e-masjid.my/api/khairat-api/repository"
)

type DependentService interface {
	Save(dependent model.Dependent, memberId int) error
	Delete(d int) error
}

type DependentServiceImpl struct {
	dependentRepository repository.DependentRepository
	memberRepository    repository.MemberRepository
	personRepository    repository.PersonRepository
}

func NewDependentService(dependentRepository repository.DependentRepository, memberRepository repository.MemberRepository, personRepository repository.PersonRepository) DependentService {
	return &DependentServiceImpl{
		dependentRepository: dependentRepository,
		memberRepository:    memberRepository,
		personRepository:    personRepository,
	}
}

// Delete implements DependentService.
func (repo *DependentServiceImpl) Delete(id int) error {
	dependent, err := repo.dependentRepository.FindById(id)

	if err != nil {
		return err
	}

	person, err := repo.personRepository.FindById(dependent.Person.Id)

	if err != nil {
		return err
	}

	err = repo.dependentRepository.Delete(dependent)

	if err != nil {
		return err
	}

	err = repo.personRepository.Delete(person.Id)

	if err != nil {
		return err
	}

	return nil
}

// Save implements DependentService.
func (repo *DependentServiceImpl) Save(dependent model.Dependent, memberId int) error {
	newPerson, err := repo.personRepository.Save(dependent.Person)
	if err != nil {
		return err
	}
	dependent.Person = newPerson

	member, err := repo.memberRepository.FindById(memberId)
	if err != nil {
		return err
	}
	dependent.Member = member

	err = repo.dependentRepository.Save(dependent, int64(memberId))
	if err != nil {
		return err
	}

	return nil
}
